### PR TITLE
Core/Quests: Abandon&Finish quest items

### PR DESCRIPTION
**Pull request by poszer**
Destroy quest items on quest failure/abandon.

Failing/abandoning a quest should not unconditionally remove required items, non-quest specific items such as crafting reagents should remain in inventory.
Remove items if the quest is abandoned.

##### CHANGES PROPOSED:

-  Updated
-  

###### ISSUES ADDRESSED:

Closes [1094](https://github.com/azerothcore/azerothcore-wotlk/issues/1094)

##### TESTS PERFORMED:

Builded without errors on Win. Tested in game, works fine.

before

after

##### HOW TO TEST THE CHANGES:

1. Accept a quest which requires collecting some item.
2. Collect 1+ of that item.
3. Abandon Quest
4. Item will not remains in your inventory.
5. Also do quests like [HeadHunting](http://www.wowhead.com/quest=582/headhunting) ->In this type of quest you can loot more than required items. At the end,  It should remove items on quest abandon/finish.
6. Or the profession quest --> Get 2 or 3 Damaged Necklace items (.add 43297) , and turn 1 in quest. Finish it, or abandon it
7. And Test it

I will write some quests, so you can test it faster
.quest add 7798
.quest add 9433
.quest add 1153
.quest add 1102

##### KNOWN ISSUES AND TODO LIST:

- [ ]
- [ ] 

##### Target branch(es):

Master

### CODE REVIEW

**Comment by wetbrownsauce** [comment]
Tested in game. 

- Quest Dwarven Outfitters correctly deleted my 8 wolf meat on abandon.
- Quest Concerted Efforts correctly left my battleground marks alone on abandon.
- Quest A dip in the moonwell correctly deleted both the provided item as well as the collected quest item on abandon. 
- Quest Elixirs for the bladeleafs correctly left my elixirs alone on abandon.
- Quest Elixirs for the bladeleafs correctly accepted the right number of elixirs even though I had more in inventory when handing in.
- Quest A donation of silk correctly left my 200 silk untouched on abandon.
- Quest Consumed by hatred correctly deleted my 120 bristleboar tusks on abandon (only 60 required for quest). 
- Quest Elixir of Agony correctly deleted the provided item, full amounts of the two required quest items, and correctly left the required non-quest item alone. 


No complaints from in-game testing unless I missed anything, if anyone wants to sign off on the code then go for it :)

**Comment by Maurowin** [comment]
To me its fine too

**Comment by BarbzYHOOL** [comment]
Ok, if no objection on code, we'll merge tomorrow or smthg!

File: src/server/game/Entities/Player/Player.cpp [hunk 1]
```diff
@@ -15985,14 +15993,32 @@ void Player::FailQuest(uint32 questId)
...
+        // It will Destroy quest items on quests abandons.
+        for (uint8 i = 0; i < QUEST_ITEM_OBJECTIVES_COUNT; ++i)
+            if (ItemTemplate const* itemTemplate = sObjectMgr->GetItemTemplate(quest->RequiredItemId[i]))
+                if (quest->RequiredItemCount[i] > 0 && itemTemplate->Bonding == BIND_QUEST_ITEM)
+                    DestroyItemCount(quest->RequiredItemId[i], 9999, true);
+
+        for (uint8 i = 0; i < QUEST_SOURCE_ITEM_IDS_COUNT; ++i)
+            if (ItemTemplate const* itemTemplate = sObjectMgr->GetItemTemplate(quest->ItemDrop[i]))
+                if (quest->ItemDropQuantity[i] > 0 && itemTemplate->Bonding == BIND_QUEST_ITEM)
+                    DestroyItemCount(quest->ItemDrop[i], 9999, true);
```
Comment (BarbzYHOOL) [inline src/server/game/Entities/Player/Player.cpp]: there is another 9999

File: src/server/game/Entities/Player/Player.cpp [hunk 2]
```diff
@@ -15746,15 +15746,23 @@ void Player::RewardQuest(Quest const* quest, uint32 reward, Object* questGiver,
...
+    }
     for (uint8 i = 0; i < QUEST_SOURCE_ITEM_IDS_COUNT; ++i)
     {
-        if (quest->ItemDrop[i])
+        if (ItemTemplate const* itemTemplate = sObjectMgr->GetItemTemplate(quest->ItemDrop[i]))
         {
-            uint32 count = quest->ItemDropQuantity[i];
-            DestroyItemCount(quest->ItemDrop[i], count ? count : 9999, true);
+            if (quest->ItemDropQuantity[i] > 0 && itemTemplate->Bonding == BIND_QUEST_ITEM && !quest->IsRepeatable() && !HasQuestForItem(quest->ItemDrop[i], quest_id))
+                DestroyItemCount(quest->ItemDrop[i], 9999, true);
```
Comment (BarbzYHOOL) [inline src/server/game/Entities/Player/Player.cpp]: 9999 but not sure if it should be changed @talamortis

File: src/server/game/Entities/Player/Player.cpp [hunk 3]
```diff
@@ -15985,14 +15993,32 @@ void Player::FailQuest(uint32 questId)
...
+                if (quest->RequiredItemCount[i] > 0 && itemTemplate->Bonding == BIND_QUEST_ITEM)
+                    DestroyItemCount(quest->RequiredItemId[i], 9999, true);
+
         for (uint8 i = 0; i < QUEST_SOURCE_ITEM_IDS_COUNT; ++i)
-            if (quest->ItemDrop[i] > 0 && quest->ItemDropQuantity[i] > 0)
-                // Destroy items received during the quest.
-                DestroyItemCount(quest->ItemDrop[i], quest->ItemDropQuantity[i], true, true);
+            if (ItemTemplate const* itemTemplate = sObjectMgr->GetItemTemplate(quest->ItemDrop[i]))
+                if (quest->ItemDropQuantity[i] > 0 && itemTemplate->Bonding == BIND_QUEST_ITEM)
+                    DestroyItemCount(quest->ItemDrop[i], 9999, true);
```
Comment (BarbzYHOOL) [inline src/server/game/Entities/Player/Player.cpp]: 9999 but not sure if it should be changed @talamortis

File: src/server/game/Entities/Player/Player.cpp (hunk 1 above)
Comment (BarbzYHOOL) [inline src/server/game/Entities/Player/Player.cpp]: 9999 but not sure if it should be changed @talamortis

**Pull request merged**